elastic/kibana
# [Obs AI Assistant] ai assistant system connector

## Summary

Creates a system connector that can call the observability ai assistant to execute actions on behalf of user. The connector is tagged as tech preview.

The connector can be triggered when an alert fires. Connector can be configured with an initial message to the assistant which generates an answer and triggers potential actions on the assistant side. The current experimental scenario is to ask the assistant to generate a report of the alert that fired (by initially providing some context in the first message), recalling any information/potential resolutions of previous occurrences stored in the knowledge base and also including other active alerts that may be related. One last step that can be asked to the assistant is to trigger an action, currently only sending the report (or any other message) to a preconfigured slack webhook is supported.

## Testing
_Note: when asked to send a message to another connector (in our case slack), we'll try to include a link to the generated conversation. It is only possible to generate this link if [server.publicBaseUrl](https://www.elastic.co/guide/en/kibana/current/settings.html#server-publicBaseUrl) is correctly set in kibana settings._

- Create a slack webhook connector
  - Get slack webhook. I can share one and invite you to the workspace, or if you want to create one:
    - create personal workspace at https://slack.com/signin#workspaces
    - create an app for that workspace at https://api.slack.com/apps
    - under Features > OAuth & Permissions > Scopes > Bot Token Scopes, add `incoming-webhook` permission
    - install the app
    - webhook url is available under Features > Incoming Webhooks
- Create a rule that can be triggered with available documents and attach observability AI assistant connector. (I use `Error Count Threshold` and generate errors via `node scripts/synthtrace many_errors.ts --live`)
  - configure the connector with one genai connector and a message with instructions. Example:
```
High error count alert has triggered. Execute the following steps:
  - create a graph of the error count for the service impacted by the alert for the last 24h
  - to help troubleshoot recall past occurrences of this alarm, also any other active alerts. Generate a report with all the found informations and send it to slack connector as a single message. Also include the link to this conversation in the report
```
- Track alert status and verify connector was executed. You should get a slack notification sent by the assistant, and a new conversation will be stored

TODO
- unit/integration tests - see https://github.com/elastic/kibana/pull/168369 for reference implementation
- documentation
